### PR TITLE
chrony: Configure chrony to start only when /dev/ptp0 exists.

### PR DESCRIPTION
**Pull request by amshinde**
Hypercall to implement virtual PTP was introduced in kernel 4.10
Have chrony run only if the device created by kvm-ptp exists.
Add this as a ConditionExists in the systemd service file.

This service if named as chrony.service in deb based distributions
rather than chronyd.service, although a systemd alias exists.
However it is not possible to come up with a generic `PATH` systemd
unit relying on the alias.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

### CODE REVIEW

File: rootfs-builder/rootfs.sh [hunk 1]
```diff
@@ -381,6 +381,13 @@ echo "refclock PHC /dev/ptp0 poll 3 dpoll -2 offset 0" >> ${chrony_conf_file}
 # Reference:  https://chrony.tuxfamily.org/doc/3.4/chrony.conf.html
 sed -i 's/^\(server \|pool \|peer \)/# &/g'  ${chrony_conf_file}
 
+chrony_systemd_service="${ROOTFS_DIR}/usr/lib/systemd/system/chronyd.service"
+if [ ${distro} == ubuntu ] || [ ${distro} == debian ] ; then
+	chrony_systemd_service="${ROOTFS_DIR}/lib/systemd/system/chrony.service"
+fi 
+
+sed -i 's/^Unit/&ConditionPathExists=/dev/ptp0/' ${chrony_systemd_service}
```
Comment (jodh-intel) [inline rootfs-builder/rootfs.sh]: Is this going to work as you have slashes in the replacement text? Maybe:

```sh
sed -i 's!^Unit!&ConditionPathExists=/dev/ptp0!' ${chrony_systemd_service}
```

Comment (amshinde) [reply rootfs-builder/rootfs.sh]: @jodh-intel Looks like I missed pushing my latest changes. I have addressed this. PTAL.

File: rootfs-builder/rootfs.sh [hunk 2]
```diff
@@ -381,6 +381,13 @@ echo "refclock PHC /dev/ptp0 poll 3 dpoll -2 offset 0" >> ${chrony_conf_file}
 # Reference:  https://chrony.tuxfamily.org/doc/3.4/chrony.conf.html
 sed -i 's/^\(server \|pool \|peer \)/# &/g'  ${chrony_conf_file}
 
+chrony_systemd_service="${ROOTFS_DIR}/usr/lib/systemd/system/chronyd.service"
+if [ ${distro} == ubuntu ] || [ ${distro} == debian ] ; then
+	chrony_systemd_service="${ROOTFS_DIR}/lib/systemd/system/chrony.service"
+fi 
+
+sed -i '/^\[Unit\]/a ConditionPathExists=\/dev\/ptp0' ${chrony_systemd_service}
```
Comment (jodh-intel) [inline rootfs-builder/rootfs.sh]: Responding to @jcvenegas's thought about creating our own unit, I think the way systemd expects use to handle this is by using a snippet. Something like:

```sh
$ sudo mkdir -p /etc/systemd/system/chronyd.service.d/
$ cat <<EOT | sudo tee /etc/systemd/system/chronyd.service.d/condition-path-exists.conf
[Unit]
ConditionPathExists=/dev/ptp0
EOT
```

Comment (jcvenegas) [reply rootfs-builder/rootfs.sh]: Nice alternative, just wonder if this could be distro agnostic better! :)

**Comment by Pennyzct** [comment]
Hi~ all @jodh-intel @amshinde 
ARM CI failed on `dmesg log test` from docker suite. Since `kvm-ptp` isn't supported on AArch64 for now, chrnoy service couldn't be working well on the guest. And I think this PR could fix this failure. ;)
@grahamwhaley Since we already have two ARM CI nodes, could we add it to the `osbuilder/` repo? ;)

**Comment by grahamwhaley** [comment]
@Pennyzct - you mean enable the ARM CI to track the PRs on the osbuilder repo - I don't see why not - we already have x86 and Power on there.
You are OK to create that new Jenkins job yourself @Pennyzct ?
/cc @chavafg @GabyCT

**Comment by Pennyzct** [comment]
I'm not sure that I have the authority to do this job and is there any tutorial?? 😢 sooooorry, I'm a really rookie in this field.

**Comment by grahamwhaley** [comment]
Hi @Pennyzct - ah, indeed, you have to be in the github kata 'jenkins-admin' team to have jenkins master superpowers.... If you would like to join the admin team, let's discuss with @chavafg 
The basic workflow in Jenkins is:
- create a new freestyle job, cloning that from the 'nearest' existing job to what you want
- go through all the dialogs and 'advanced' dialogs and text/script boxes of the new job config and basically doing `s/old job stuff/new job stuff/'. For instance, we would probably copy the proxy ARM job, and then change all occurrences of 'proxy' in the dialogs to 'osbuilder'.

There are a few special corner cases, such as all jobs probably mention the test repo at some point in their configuration, and when configuring a job *for* the test repo you need to stare harder to make sure they are all correct :-) But mostly it is a copy/paste/edit of an existing job.

Let me go make an ARM osbuilder job for you...

**Comment by grahamwhaley** [comment]
OK @Pennyzct - job created: http://jenkins.katacontainers.io/job/kata-containers-osbuilder-ARM-18.04-PR/
Looks like I can't kick off a test build, as it needs a PR number - so we can either wait for the next PR, or we could create a test PR if we really needed.

**Comment by grahamwhaley** [comment]
and then one triggered anyway - let's see how this goes @Pennyzct http://jenkins.katacontainers.io/job/kata-containers-osbuilder-ARM-18.04-PR/2/console

**Comment by chavafg** [comment]
@amshinde can you add a `Fixes: #...` to the commit message so that the CI can properly run?

**Comment by Pennyzct** [comment]
@grahamwhaley thanks for the detailed instruction and the created ARM osbuilder job. ;). happy to have the super power, 嘿嘿🤩.

**Comment by egernst** [comment]
CI appears to hate your PR, @amshinde

**Comment by amshinde** [comment]
@egernst Yes, all of them seem to be failing for different reasons.
Retesting, before I go look what the issue is.

**Comment by Pennyzct** [comment]
Hi~ @amshinde 
ARM keeps failing on the same spot. ;)
```
ERROR: Failed to run command /usr/bin/git rev-list --no-merges --reverse origin/master..9c97548cb3379ecafa6e62d08799e984a55dc632: exit status 128 (stdout: , stderr: fatal: Invalid revision range origin/master..9c97548cb3379ecafa6e62d08799e984a55dc632
)
```
Does it need to be re-based? @grahamwhaley

**Comment by grahamwhaley** [comment]
@Pennyzct hmm, not sure. Only thing I can think of maybe is if the ARM jenkins slave is not clearing out its workspace, or is using a GOPATH/git repo that is outside of the jenkins WORKSPACE path, so maybe it is trying to pull on an old tree or something??

Any chance you can try to reproduce by hand @Pennyzct ?

**Comment by amshinde** [comment]
@jcvenegas @chavafg I need your help here. I am seeing this error in the CI: 
```
INFO: Created summary file '/var/lib/osbuilder/osbuilder.yaml' inside rootfs
+ exit 0
touch /tmp/osbuilder-test.xVdDrfd/rootfs-osbuilder/.debian_rootfs.done
+ die 'Background build job failed'
+ msg='Background build job failed'
+ echo 'ERROR: Background build job failed'
ERROR: Background build job failed
+ exit 1
+ exit_handler
+ '[' 1 -eq 0 ']'
+ info 'ERROR: test failed'
+ s='ERROR: test failed'
+ echo -e 'INFO: ERROR: test failed\n'
INFO: ERROR: test failed
```

When I run the ci locally as `.ci/run.sh`, I am consistently seeing this behaviour:
```
[OK] Agent installed                                                                                                                                                                                        
INFO: Check init is installed                                                                                                                                                                               
[OK] init is installed                                                                                                                                                                                      
INFO: Creating summary file                                                                                                                                                                                 
INFO: Created summary file '/var/lib/osbuilder/osbuilder.yaml' inside rootfs                                                                                                                                
Sending build context to Docker daemon  41.47kB                                                                                                                                                             
Step 1/3 : From fedora:latest                                                                                                                                                                               
latest: Pulling from library/fedora                                                                                                                                                                         
8f6ac7ed4a91: Pull complete                                                                                                                                                                                 
Digest: sha256:9c78c69f748953ba8fdb6eb9982e1abefe281d9b931a13f251eb8aec988353de                                                                                                                             
Status: Downloaded newer image for fedora:latest                                                                                                                                                            
 ---> 289289d1a15b                                                                                                                                                                                          
Step 2/3 : RUN [ -n "$http_proxy" ] && sed -i '$ a proxy='$http_proxy /etc/dnf/dnf.conf ; true                                                                                                              
 ---> Running in db4c239044c6                                                                                                                                                                               
Removing intermediate container db4c239044c6                                                                                                                                                                
 ---> 75ba7bfe3e18                                                                                                                                                                                          
Step 3/3 : RUN dnf install -y qemu-img parted gdisk e2fsprogs gcc xfsprogs findutils                                                                                                                        
 ---> Running in ac7c695a193e                                                                                                                                                                               
Fedora Modular 30 - x86_64                      1.5 MB/s | 2.7 MB     00:01                                                                                                                                 
Fedora Modular 30 - x86_64 - Updates            2.7 MB/s | 2.0 MB     00:00                                                                                                                                 
Fedora 30 - x86_64 - Updates                    7.3 MB/s |  14 MB     00:01                                                                                                                                 
Fedora 30 - x86_64                               16 kB/s | 3.4 kB     00:00                                                                                                                                 
Failed to synchronize cache for repo 'fedora'                                                                                                                                                               
Error: Failed to synchronize cache for repo 'fedora'                                                                                                                                                        
The command '/bin/sh -c dnf install -y qemu-img parted gdisk e2fsprogs gcc xfsprogs findutils' returned a non-zero code: 1                                                                                  
Failed at 121: sudo -E AGENT_INIT="${AGENT_INIT}" USE_DOCKER=true ./image-builder/image_builder.sh "$ROOTFS_DIR"                                                                                            
Failed at 18: "${cidir}/install_kata_image.sh"                                                                                                                                                              
+ exit_handler                                                                                                                                                                                              
+ '[' 1 -eq 0 ']'                                                                                                                                                                                           
+ info 'ERROR: test failed'                                                                                                                                                                                 
+ s='ERROR: test failed'                                                                                                                                                                                    
+ echo -e 'INFO: ERROR: test failed\n'                                                                                                                                                                      
INFO: ERROR: test failed                                                                                                                                                                                    
                                                                                                                                                                                                            
+ '[' -d /tmp/osbuilder-test.6OINQxE/rootfs-osbuilder ']'                                                                                                                                                   
+ info 'no rootfs created'                                                                                                                                                                                  
+ s='no rootfs created'                                                                                                                                                                                     
+ echo -e 'INFO: no rootfs created\n'                                                                                                                                                                       
INFO: no rootfs created 
```

The failure being: `Failed to synchronize cache for repo 'fedora'  `
Can you reproduce this? The CI does not give any clue of what failed.

**Comment by chavafg** [comment]
@amshinde trying to reproduce... will get back when I have something.

The CI failure means that the `rootfses` were not built correctly, but I cannot see exactly were they failed.

**Comment by chavafg** [comment]
for some reason the travis jobs are hitting the 50m timeout... Although I see that also the travis master branch job has the same issue. 
@grahamwhaley I went and fix the configuration of the ARM job because we do not use the `jenkins_job_build.sh` as we have other flow for this repository.

**Comment by Pennyzct** [comment]
Hi~ @chavafg @grahamwhaley I saw that ARM CI and Ubuntu-16-04 has fallen into the same error:
```
chronic /tmp/jenkins/workspace/kata-containers-osbuilder-ubuntu-16-04-PR/go/src/github.com/kata-containers/osbuilder/tests/../../tests/cmd/kata-manager/kata-manager.sh configure-image
Build step 'Execute shell' marked build as failure
```
any thoughts??? 😵😵😵

**Comment by grahamwhaley** [comment]
@chavafg - ah, thanks for the job fix :-)
@Pennyzct I dunno - it's not just ARM though is it? - let's see what @chavafg @amshinde come up with - unless you have time to do a hand run locally and try to capture more details ... ;-)

**Comment by jodh-intel** [comment]
I think the Travis job is failing here (but don't know why):

```
/home/travis/gopath/src/github.com/kata-containers/agent /
+ make install DESTDIR=/rootfs INIT=no SECCOMP=
rm -f kata-agent kata-agent.service
+ make INIT=no
install -D kata-agent /rootfs/usr/bin/kata-agent
install: cannot stat 'kata-agent': No such file or directory
```

**Comment by jodh-intel** [comment]
Buried in the proxy log artifact in Jenkins, I found this:

```
[    1.069551] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc2.d, not generating chrony.service.
[    1.070131] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc3.d, not generating chrony.service.
[    1.070516] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc4.d, not generating chrony.service.
[    1.070889] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc5.d, not generating chrony.service.
[\x1b[0;1;31m!!!!!!\x1b[0m] Failed to isolate default target, freezing.
```

/cc @amshinde

**Comment by jodh-intel** [comment]
@amshinde - any progress on debugging this issue?

**Comment by amshinde** [comment]
@jodh-intel Thanks for your pointer. I havent had the chance to look a  great deal at this yesterday. I'll try to debug this today. 
Whats concerning is, I see a similar failure on a test PR that I raised a couple of days back.
https://github.com/kata-containers/osbuilder/pull/315

So, it looks like this is failing on the master as well.

**Comment by marcov** [comment]
> Buried in the proxy log artifact in Jenkins, I found this:
> 
> ```
> [    1.069551] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc2.d, not generating chrony.service.
> [    1.070131] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc3.d, not generating chrony.service.
> [    1.070516] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc4.d, not generating chrony.service.
> [    1.070889] systemd-sysv-generator[69]: Ignoring S01chrony symlink in rc5.d, not generating chrony.service.
> [\x1b[0;1;31m!!!!!!\x1b[0m] Failed to isolate default target, freezing.
> ```
> 
> /cc @amshinde

@amshinde in case you missed, #319 should be the fix for this.

**Comment by amshinde** [comment]
> @amshinde in case you missed, #319 should be the fix for this.

Thanks @marcov! Thats one thing out of my todo today :)
Thanks for fixing it. I'll rebase once your PR is merged.
